ui/desktop/src/components/MCPUIResourceRenderer.tsx

```diff
@@ -2,64 +2,157 @@ import { UIResourceRenderer, UIActionResult } from '@mcp-ui/client';
 
-      default: {
-        console.warn('unsupported message sent from MCP-UI:', result);
-        break;
+            const toolName =
+              getString(toolData, 'toolName') ||
+              getString(toolData, 'tool') ||
+              getString(toolData, 'name');
+            // MCP UI sends 'params' not 'arguments'
+            const toolArguments =
+              getRecord(toolData, 'params') ||
+              getRecord(toolData, 'parameters') ||
+              getRecord(toolData, 'arguments') ||
+              {};
+
+            if (!toolName) {
+              throw new Error('Tool name not found in action result');
+            }
+
+            const requestBody = {
+              tool_name: toolName,
+              arguments: toolArguments,
+              session_id: null, // You may want to pass the current session ID if available
+            };
+
+            const response = await fetch(getApiUrl('/agent/execute_tool'), {
+              method: 'POST',
+              headers: {
+                'Content-Type': 'application/json',
+                'X-Secret-Key': await window.electron.getSecretKey(),
+              },
+              body: JSON.stringify(requestBody),
+            });
+
+            if (!response.ok) {
+              const errorText = await response.text();
+              throw new Error(`Tool execution failed: ${errorText}`);
+            }
+
+            const data = await response.json();
+
+            if (data.success) {
+              // Emit event so chat can display tool request/response
+              try {
+                window.dispatchEvent(
+                  new CustomEvent('external-tool-executed', {
+                    detail: { toolName, toolArguments, result: data.result },
+                  })
+                );
+                // Request chat to scroll to bottom to reveal new messages
+                window.dispatchEvent(new CustomEvent('scroll-chat-to-bottom'));
+              } catch (e) {
+                // ignore event dispatch errors
+              }
+              handleAction(result, false);
+            } else {
+              throw new Error(data.error || 'Tool execution failed');
+            }
+          } catch (error) {
+            toast.error(
+              `Failed to execute tool: ${error instanceof Error ? error.message : String(error)}`
+            );
+            console.error(error);
+          }
+          break;
+        }
+
+        default: {
+          console.warn('unsupported message sent from MCP-UI:', result);
+          break;
+        }
       }
-    }
-  }, []);
+    },
+    [getProp, getRecord, getString]
+  );
 
   return (
-    <div className="mt-3 p-4 border border-borderSubtle rounded-lg bg-background-muted">
+    <div className="mt-3 p-1 border border-borderSubtle rounded-lg bg-background-muted">
       <div className="overflow-hidden rounded-sm">
         <UIResourceRenderer
           resource={content.resource}
```

Comment:
Adding a `style` here can help adjust the iframes height, where `ui-size-change` is not thrown or the UI is not responsive:
```tsx
  <UIResourceRenderer
    resource={content.resource}
              onUIAction={handleUIAction}
    htmlProps={{
      style: { height: 400px, minHeight: 0 }
      autoResizeIframe: {
        height: true,
        width: false,
      },
    }}
  />
```